content/blog/fargate-vs-ec2/index.md

```diff
@@ -0,0 +1,167 @@
+
+```diff
+eks_cluster = eks.Cluster("eks-cluster",
+    vpc_id=eks_vpc.vpc_id,
+    private_subnet_ids=eks_vpc.private_subnet_ids,
+-   fargate=True,  
++   instance_type="t3.medium",
++   desired_capacity=2,
++   min_size=1,
++   max_size=2,
+)
+```
+
+Now there are only two nodes, with the pods distributed among them. Those are my two `t3.medium`s.
+
+![k9s eks demo](k9s-eks.gif)
+
+So when should you use Fargate, and when EC2 for EKS? Well, it really all relates to what we are seeing here with the nodes.
+
+Let me explain.
+
+## Bin Packing
+
+Normally, the Kubernetes scheduler has to solve a bin-packing problem—fitting M pods across N nodes based on available resources. This is a straight-up resource optimization problem of balancing CPU, memory, and other resource demands across the cluster.
+
+This can be surprisingly complex in practice. Teams must balance selecting the right instance types for different workload needs while managing multiple node groups. Adding to this complexity are scheduling rules and resource quotas that need careful configuration. Perhaps most challenging is striking the right balance between high resource utilization and maintaining enough headroom for spikes in demand.
+
+![ec2 pods](ec2-pods.png)
+
+With Fargate, AWS sidesteps this challenge by providing a "just-in-time" bin for each pod. Each Fargate pod runs on its own dynamically provisioned, right-sized mini-environment, where the "bin" (the Fargate instance) is exactly sized to match the pod's requested resources. This means there's no need for Kubernetes to optimize resource usage across a pool of shared nodes, as each pod effectively has its own "container" provided by Fargate that fits it precisely.
+
+In other words, AWS effectively pushes this bin-packing responsibility to itself, so you don't have to worry ( as much ) about it.
+
+![Fargate pods](faregate-pods.png)
+
+{{% notes type="info" %}}
+While Fargate abstracts away node management, it's important to understand that Fargate pods [still run on EC2 instances](https://justingarrison.com/blog/2024-02-08-fargate-is-not-firecracker/) behind the scenes. And each Fargate pod gets its own ENI (Elastic Network Interface) that can sometimes limit your scaling because pods aren't sharing a network namespace.
+{{% /notes %}}
+
+## Pros and Cons
+
+So yeah, Fargate abstracts away the complexity of managing nodes and scaling clusters by shifting bin-packing to AWS. With its extensive infrastructure, AWS is arguably better positioned to optimize placements at scale. They can monitor their entire cloud landscape, placing Fargate instances dynamically across available resources. This allows them to potentially leverage underutilized resources in ways that would be challenging or impractical at an individual account's scale.
+
+Ideally, this frees you to focus on the services rather than on the underlying infrastructure but in practise it's a bit more complex.
```

Comment:
```suggestion
Ideally, this frees you to focus on the services rather than on the underlying infrastructure but in practice it's a bit more complex.
```